## Coursera Applied Data Science Capstone 
### Week 3 Assignment: part I II and III
#### Segmenting and Clustering Neighborhoods in Toronto

## Part 1

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 

So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.

Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

## Web Scraping
Here I use pandas read method to scrape the webpage.

In [1]:
import pandas as pd
import numpy as np

In [2]:
!pip install BeautifulSoup4

In [3]:
import html5lib

In [4]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df=pd.read_html(url, flavor='html5lib')[0]

In [5]:
df

Postcode           Borough          Neighbourhood
0        M1A      Not assigned           Not assigned
1        M2A      Not assigned           Not assigned
2        M3A        North York              Parkwoods
3        M4A        North York       Victoria Village
4        M5A  Downtown Toronto           Harbourfront
..       ...               ...                    ...
283      M8Z         Etobicoke              Mimico NW
284      M8Z         Etobicoke     The Queensway West
285      M8Z         Etobicoke  Royal York South West
286      M8Z         Etobicoke         South of Bloor
287      M9Z      Not assigned           Not assigned

[288 rows x 3 columns]

In [6]:
df.shape

(288, 3)

The dataframe consists of 3 columns: PostalCode, Borough, and Neighborhood
It has 288 entries.

## Data Cleaning

I Only process the cells that have an assigned borough. i.e. Ignore cells with a borough that is Not assigned.

In [7]:
df=df[df.Borough != 'Not assigned']
df

Postcode           Borough             Neighbourhood
2        M3A        North York                 Parkwoods
3        M4A        North York          Victoria Village
4        M5A  Downtown Toronto              Harbourfront
5        M5A  Downtown Toronto               Regent Park
6        M6A        North York          Lawrence Heights
..       ...               ...                       ...
282      M8Z         Etobicoke  Kingsway Park South West
283      M8Z         Etobicoke                 Mimico NW
284      M8Z         Etobicoke        The Queensway West
285      M8Z         Etobicoke     Royal York South West
286      M8Z         Etobicoke            South of Bloor

[211 rows x 3 columns]

77 entries were excluded under this criteria.


More than one neighborhood can exist in one postal code area. 
For example, in the table on the Wikipedia page,  M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows need to be combined into one row with the neighborhoods separated with a comma.

Here I group each postcode by 'Borough' and  get the corresponding neighbourhoods as a list by using apply(list).

In [8]:
df=df.groupby(['Postcode','Borough'])['Neighbourhood'].apply(list)
df

Postcode  Borough    
M1B       Scarborough                                     [Rouge, Malvern]
M1C       Scarborough             [Highland Creek, Rouge Hill, Port Union]
M1E       Scarborough                  [Guildwood, Morningside, West Hill]
M1G       Scarborough                                             [Woburn]
M1H       Scarborough                                          [Cedarbrae]
                                               ...                        
M9N       York                                                    [Weston]
M9P       Etobicoke                                            [Westmount]
M9R       Etobicoke      [Kingsview Village, Martin Grove Gardens, Rich...
M9V       Etobicoke      [Albion Gardens, Beaumond Heights, Humbergate,...
M9W       Etobicoke                                            [Northwest]
Name: Neighbourhood, Length: 103, dtype: object

Join neighbourhoods in the same borough and included in a dataframe and reset the index.

In [9]:
df=df.apply(lambda x: ','.join(x)).to_frame().reset_index()
df

Postcode      Borough                                      Neighbourhood
0        M1B  Scarborough                                      Rouge,Malvern
1        M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough                    Guildwood,Morningside,West Hill
3        M1G  Scarborough                                             Woburn
4        M1H  Scarborough                                          Cedarbrae
..       ...          ...                                                ...
98       M9N         York                                             Weston
99       M9P    Etobicoke                                          Westmount
100      M9R    Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
101      M9V    Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...
102      M9W    Etobicoke                                          Northwest

[103 rows x 3 columns]

If a cell has a borough but No assigned neighborhood, then the neighborhood will be the same as the borough.

In [10]:
for index,row in df.iterrows():
    if row['Neighbourhood']=='Not assigned':
        row['Neighbourhood']=row['Borough']

#### here is the Final output for part I

In [11]:
df.head(20)

Postcode      Borough                                      Neighbourhood
0       M1B  Scarborough                                      Rouge,Malvern
1       M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2       M1E  Scarborough                    Guildwood,Morningside,West Hill
3       M1G  Scarborough                                             Woburn
4       M1H  Scarborough                                          Cedarbrae
5       M1J  Scarborough                                Scarborough Village
6       M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park
7       M1L  Scarborough                      Clairlea,Golden Mile,Oakridge
8       M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West
9       M1N  Scarborough                         Birch Cliff,Cliffside West
10      M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
11      M1R  Scarborough                                   Maryvale,Wexford
12      M1S  Scarborough                                          Agincourt
13      M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
14      M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...
15      M1W  Scarborough                                    L'Amoreaux West
16      M1X  Scarborough                                        Upper Rouge
17      M2H   North York                                  Hillcrest Village
18      M2J   North York                         Fairview,Henry Farm,Oriole
19      M2K   North York                                    Bayview Village

In [12]:
print('Data frame size =', df.shape)

Data frame size = (103, 3)


# Part 2

In [13]:
df1=pd.read_csv('http://cocl.us/Geospatial_data') # read csv file to a data frame.
df1

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [14]:
df1.rename(columns={'Postal Code':'Postcode'},inplace=True) # change the column name 'Postal Code' to 'Postcode' to match with df dataframe
df1.set_index('Postcode') # set index to 'Postcode': This lets merge dataframes later withrespect to indices

Latitude  Longitude
Postcode                      
M1B       43.806686 -79.194353
M1C       43.784535 -79.160497
M1E       43.763573 -79.188711
M1G       43.770992 -79.216917
M1H       43.773136 -79.239476
...             ...        ...
M9N       43.706876 -79.518188
M9P       43.696319 -79.532242
M9R       43.688905 -79.554724
M9V       43.739416 -79.588437
M9W       43.706748 -79.594054

[103 rows x 2 columns]

In [15]:
df.set_index('Postcode') # set index to 'Postcode' in df dataframe

Borough                                      Neighbourhood
Postcode                                                                
M1B       Scarborough                                      Rouge,Malvern
M1C       Scarborough               Highland Creek,Rouge Hill,Port Union
M1E       Scarborough                    Guildwood,Morningside,West Hill
M1G       Scarborough                                             Woburn
M1H       Scarborough                                          Cedarbrae
...               ...                                                ...
M9N              York                                             Weston
M9P         Etobicoke                                          Westmount
M9R         Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...
M9V         Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...
M9W         Etobicoke                                          Northwest

[103 rows x 2 columns]

In [16]:
Neigh_Toronto=pd.merge(df,df1) # merge df and df1 dataframes. They both have the same index column
Neigh_Toronto

Postcode      Borough                                      Neighbourhood  \
0        M1B  Scarborough                                      Rouge,Malvern   
1        M1C  Scarborough               Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough                    Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                             Woburn   
4        M1H  Scarborough                                          Cedarbrae   
..       ...          ...                                                ...   
98       M9N         York                                             Weston   
99       M9P    Etobicoke                                          Westmount   
100      M9R    Etobicoke  Kingsview Village,Martin Grove Gardens,Richvie...   
101      M9V    Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...   
102      M9W    Etobicoke                                          Northwest   

      Latitude  Longitude  
0    43.806686 -79.194353  
1    43.784535 -79.160497  
2    43.763573 -79.188711  
3    43.770992 -79.216917  
4    43.773136 -79.239476  
..         ...        ...  
98   43.706876 -79.518188  
99   43.696319 -79.532242  
100  43.688905 -79.554724  
101  43.739416 -79.588437  
102  43.706748 -79.594054  

[103 rows x 5 columns]

In [17]:
# let's explore bouroughs
Neigh_Toronto['Borough'].unique()

array(['Scarborough', 'North York', 'East York', 'East Toronto',
       'Central Toronto', 'Downtown Toronto', 'York', 'West Toronto',
       "Queen's Park", 'Mississauga', 'Etobicoke'], dtype=object)

### Use geopy library to get the latitude and longitude values of Toronto.

In [18]:
!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    geographiclib-1.49         |             py_0          32 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         237 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0        conda-forge
    geopy:         1.20.0-py_0      conda-forge

The following pac

In [19]:
address = 'Toronto, ON, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))        

The geograpical coordinate of Toronto are 43.653963, -79.387207.


## Toronto map

In [20]:
!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.



In [21]:
# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(Neigh_Toronto['Latitude'], Neigh_Toronto['Longitude'], Neigh_Toronto['Borough'], Neigh_Toronto['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

### Foursquare API to explore the neighborhoods and segment them.

In [22]:
CLIENT_ID = 'KLXCYCV12KIEPO2PD2XLFYZY5EVEBVUY25RO5VFNW5EA05X4' # Foursquare ID
CLIENT_SECRET = 'X5Y5AKIRBCB3SWOZFTBZMSKUBQ1ZON1XLAOBXZQZI3RWYEC4' # Foursquare Secret
VERSION = '20180605' # Foursquare API version

### Explore neighbourhoods in Toronto

Let's first explore the last neighbourhood in Toronto. (Our goal is to examine one neighbourhood. that is the main reason I selected this last neighbourhood because it contain only one neighbourhood unlike the first.

In [23]:
# get the neighbourhood name
Neigh_Toronto.loc[102, 'Neighbourhood']

'Northwest'

In [24]:
neighborhood_latitude = Neigh_Toronto.loc[102, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = Neigh_Toronto.loc[102, 'Longitude'] # neighborhood longitude value
neighborhood_name = Neigh_Toronto.loc[102,'Neighbourhood']#neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Northwest are 43.706748299999994, -79.5940544.


In [25]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [26]:
limit=100
radius=500
url_NW = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    limit)

url_NW



'https://api.foursquare.com/v2/venues/explore?&client_id=KLXCYCV12KIEPO2PD2XLFYZY5EVEBVUY25RO5VFNW5EA05X4&client_secret=X5Y5AKIRBCB3SWOZFTBZMSKUBQ1ZON1XLAOBXZQZI3RWYEC4&v=20180605&ll=43.706748299999994,-79.5940544&radius=500&limit=100'

In [27]:
results_NW = requests.get(url_NW).json()
results_NW

{'meta': {'code': 200, 'requestId': '5d7fe4a8be70780039255d66'},
 'response': {'headerLocation': 'Toronto',
  'headerFullLocation': 'Toronto',
  'headerLocationGranularity': 'city',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.711248304499996,
    'lng': -79.58784096674265},
   'sw': {'lat': 43.70224829549999, 'lng': -79.60026783325736}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '55dc6273498eea1c592572ef',
       'name': 'Up Express On Board',
       'location': {'lat': 43.70508925853441,
        'lng': -79.59780379275085,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.70508925853441,
          'lng': -79.59780379275085}],
        'distance': 353,
        'cc': 'CA',
        'country': 'Canada',
        'formattedAddress': ['Cana

all the information is in the *items* key

In [28]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Let's clean the json and structure it into a *pandas* dataframe.

In [29]:
venues = results_NW['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0     Up Express On Board        Moving Target  43.705089 -79.597804
1      Economy Rent A Car  Rental Car Location  43.708471 -79.589943
2  Logistics Distribution                  Bar  43.707554 -79.589252
3           Saand Rexdale            Drugstore  43.705072 -79.598725

### Now let's do the above to all neighborhoods in Toronto.

Following is a function to get information about 100 venues within 500 radius around Toronto

In [30]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, limit=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [31]:
Toronto_venues = getNearbyVenues(names=Neigh_Toronto['Neighbourhood'],latitudes=Neigh_Toronto['Latitude'],
                                   longitudes=Neigh_Toronto['Longitude'])

In [32]:
Toronto_venues

Neighbourhood  \
0                                         Rouge,Malvern   
1                  Highland Creek,Rouge Hill,Port Union   
2                       Guildwood,Morningside,West Hill   
3                       Guildwood,Morningside,West Hill   
4                       Guildwood,Morningside,West Hill   
...                                                 ...   
2249  Albion Gardens,Beaumond Heights,Humbergate,Jam...   
2250                                          Northwest   
2251                                          Northwest   
2252                                          Northwest   
2253                                          Northwest   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.806686               -79.194353   
1                  43.784535               -79.160497   
2                  43.763573               -79.188711   
3                  43.763573               -79.188711   
4                  43.763573               -79.188711   
...                      ...                      ...   
2249               43.739416               -79.588437   
2250               43.706748               -79.594054   
2251               43.706748               -79.594054   
2252               43.706748               -79.594054   
2253               43.706748               -79.594054   

                                Venue  Venue Latitude  Venue Longitude  \
0                             Wendy's       43.807448       -79.199056   
1               Royal Canadian Legion       43.782533       -79.163085   
2     Swiss Chalet Rotisserie & Grill       43.767697       -79.189914   
3                   G & G Electronics       43.765309       -79.191537   
4                          Marina Spa       43.766000       -79.191000   
...                               ...             ...              ...   
2249                      Tim Hortons       43.742015       -79.589690   
2250              Up Express On Board       43.705089       -79.597804   
2251               Economy Rent A Car       43.708471       -79.589943   
2252           Logistics Distribution       43.707554       -79.589252   
2253                    Saand Rexdale       43.705072       -79.598725   

            Venue Category  
0     Fast Food Restaurant  
1                      Bar  
2              Pizza Place  
3        Electronics Store  
4                      Spa  
...                    ...  
2249           Coffee Shop  
2250         Moving Target  
2251   Rental Car Location  
2252                   Bar  
2253             Drugstore  

[2254 rows x 7 columns]

In [33]:
print(Toronto_venues.shape)

(2254, 7)


In [34]:
Toronto_venues.groupby('Neighbourhood').count() # get the number of different venue catogories in each neighbourhood. 
                            # Please note: above getNearbyVenues function we assigned search limit to 100 and radius to 500. 

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                 100   
Agincourt                                                                4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                       3   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                       9   
Alderwood,Long Branch                                                   10   
...                                                                    ...   
Willowdale West                                                          7   
Woburn                                                                   4   
Woodbine Gardens,Parkview Hill                                          12   
Woodbine Heights                                                         8   
York Mills West                                                          3   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide,King,Richmond                                                  100   
Agincourt                                                                 4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                        3   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                        9   
Alderwood,Long Branch                                                    10   
...                                                                     ...   
Willowdale West                                                           7   
Woburn                                                                    4   
Woodbine Gardens,Parkview Hill                                           12   
Woodbine Heights                                                          8   
York Mills West                                                           3   

                                                    Venue  Venue Latitude  \
Neighbourhood                                                               
Adelaide,King,Richmond                                100             100   
Agincourt                                               4               4   
Agincourt North,L'Amoreaux East,Milliken,Steele...      3               3   
Albion Gardens,Beaumond Heights,Humbergate,Jame...      9               9   
Alderwood,Long Branch                                  10              10   
...                                                   ...             ...   
Willowdale West                                         7               7   
Woburn                                                  4               4   
Woodbine Gardens,Parkview Hill                         12              12   
Woodbine Heights                                        8               8   
York Mills West                                         3               3   

                                                    Venue Longitude  \
Neighbourhood                                                         
Adelaide,King,Richmond                                          100   
Agincourt                                                         4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                3   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                9   
Alderwood,Long Branch                                            10   
...                                                             ...   
Willowdale West                                                   7   
Woburn                                                            4   
Woodbine Gardens,Parkview Hill                                   12   
Woodbine Heights                                                  8   
York Mills West                                                   3   

                                                    Venue Category  
Neighbourhood 

In [60]:
# The number of venue catogeries
len(Toronto_venues["Venue Category"].unique())

277

There are 274 venue categories. Let's select category  as the variable to perform one hot encoding.


In [61]:
# Assigned dummies to venue catogories and concaternate with Neighbourhood column
Toronto_cat=pd.concat([Toronto_venues['Neighbourhood'],pd.get_dummies(Toronto_venues['Venue Category'])],axis=1)
Toronto_cat

Neighbourhood  Accessories Store  \
0                                         Rouge,Malvern                  0   
1                  Highland Creek,Rouge Hill,Port Union                  0   
2                       Guildwood,Morningside,West Hill                  0   
3                       Guildwood,Morningside,West Hill                  0   
4                       Guildwood,Morningside,West Hill                  0   
...                                                 ...                ...   
2249  Albion Gardens,Beaumond Heights,Humbergate,Jam...                  0   
2250                                          Northwest                  0   
2251                                          Northwest                  0   
2252                                          Northwest                  0   
2253                                          Northwest                  0   

      Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
2249                  0        0                   0             0   
2250                  0        0                   0             0   
2251                  0        0                   0             0   
2252                  0        0                   0             0   
2253                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
2249               0                0                 0                    0   
2250               0                0                 0                    0   
2251               0                0                 0                    0   
2252               0                0                 0                    0   
2253               0                0                 0                    0   

      ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0     ...              0                              0                 0   
1     ...              0                              0                 0   
2     ...              0                              0                 0   
3     ...              0                              0                 0   
4     ...              0                              0                 0   
...   ...            ...                            ...               ...   
2249  ...              0                              0                 0   
2250  ...              0                              0                 0   
2251  ...              0                              0                 0   
2252  ...              0                              0                 0   
2253  ...              0                              0                 0   

      Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0               0                      0                0         0   
1               0                      0                0         0   
2               0                      0                0         0   
3               0                      0                0         0   
4               0      

Then, let's group rows by neighborhood 

In [62]:
Toronto_grouped=Toronto_cat.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped

Neighbourhood  Accessories Store  \
0                              Adelaide,King,Richmond                0.0   
1                                           Agincourt                0.0   
2   Agincourt North,L'Amoreaux East,Milliken,Steel...                0.0   
3   Albion Gardens,Beaumond Heights,Humbergate,Jam...                0.0   
4                               Alderwood,Long Branch                0.0   
..                                                ...                ...   
94                                    Willowdale West                0.0   
95                                             Woburn                0.0   
96                     Woodbine Gardens,Parkview Hill                0.0   
97                                   Woodbine Heights                0.0   
98                                    York Mills West                0.0   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                 0.0      0.0                 0.0           0.0   
1                 0.0      0.0                 0.0           0.0   
2                 0.0      0.0                 0.0           0.0   
3                 0.0      0.0                 0.0           0.0   
4                 0.0      0.0                 0.0           0.0   
..                ...      ...                 ...           ...   
94                0.0      0.0                 0.0           0.0   
95                0.0      0.0                 0.0           0.0   
96                0.0      0.0                 0.0           0.0   
97                0.0      0.0                 0.0           0.0   
98                0.0      0.0                 0.0           0.0   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0              0.0              0.0               0.0                 0.03   
1              0.0              0.0               0.0                 0.00   
2              0.0              0.0               0.0                 0.00   
3              0.0              0.0               0.0                 0.00   
4              0.0              0.0               0.0                 0.00   
..             ...              ...               ...                  ...   
94             0.0              0.0               0.0                 0.00   
95             0.0              0.0               0.0                 0.00   
96             0.0              0.0               0.0                 0.00   
97             0.0              0.0               0.0                 0.00   
98             0.0              0.0               0.0                 0.00   

    ...  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0   ...            0.0                           0.02               0.0   
1   ...            0.0                           0.00               0.0   
2   ...            0.0                           0.00               0.0   
3   ...            0.0                           0.00               0.0   
4   ...            0.0                           0.00               0.0   
..  ...            ...                            ...               ...   
94  ...            0.0                           0.00               0.0   
95  ...            0.0                           0.00               0.0   
96  ...            0.0                           0.00               0.0   
97  ...            0.0                           0.00               0.0   
98  ...            0.0                           0.00               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0         0.000                    0.0              0.0      0.01   
1         0.000                    0.0              0.0      0.00   
2         0.000                    0.0              0.0      0.00   
3         0.000                    0.0              0.0      0.00   
4         0.000                    0.0              0.0      0.00   
..          ...                    ...              ...   

It is a has 100 rows and 275 columns

In [63]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [64]:
num_top_venues = 5

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                  Park   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1                Lounge          Skating Rink        Clothing Store   
2            Playground           Yoga Studio     Electronics Store   
3              Pharmacy            Beer Store  Fast Food Restaurant   
4              Pharmacy                   Gym          Skating Rink   

  5th Most Common Venue  
0   American Restaurant  
1           Yoga Studio  
2               Dog Run  
3        Sandwich Place  
4                   Pub

In [65]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [66]:
# set number of clusters
kclusters = 3

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0::] 

array([0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1], dtype=int32)

In [67]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

#Toronto_merged = Neigh_Toronto

# merge neighborhoods_venues_sorted with Neigh_Toronto to add latitude/longitude for each neighborhood
Toronto_merged = Neigh_Toronto.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

Toronto_merged.head() # check the last columns!

Postcode      Borough                         Neighbourhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant           Yoga Studio   
1 -79.160497             0.0                   Bar           Yoga Studio   
2 -79.188711             0.0           Pizza Place                   Spa   
3 -79.216917             0.0           Coffee Shop     Indian Restaurant   
4 -79.239476             0.0                Bakery                  Bank   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  
0     Electronics Store               Dog Run      Doner Restaurant  
1   Empanada Restaurant            Donut Shop             Drugstore  
2        Breakfast Spot   Rental Car Location        Medical Center  
3     Korean Restaurant           Yoga Studio     Electronics Store  
4      Hakka Restaurant    Athletics & Sports       Thai Restaurant

In [68]:
Toronto_merged=Toronto_merged.dropna()

In [69]:
Toronto_merged['Cluster Labels']=Toronto_merged['Cluster Labels'].astype(int)

In [70]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [72]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Cluster 1 :Grocery/Dinning

In [74]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0]

Postcode      Borough                                      Neighbourhood  \
0        M1B  Scarborough                                      Rouge,Malvern   
1        M1C  Scarborough               Highland Creek,Rouge Hill,Port Union   
2        M1E  Scarborough                    Guildwood,Morningside,West Hill   
3        M1G  Scarborough                                             Woburn   
4        M1H  Scarborough                                          Cedarbrae   
..       ...          ...                                                ...   
95       M9C    Etobicoke  Bloordale Gardens,Eringate,Markland Wood,Old B...   
96       M9L   North York                                      Humber Summit   
99       M9P    Etobicoke                                          Westmount   
101      M9V    Etobicoke  Albion Gardens,Beaumond Heights,Humbergate,Jam...   
102      M9W    Etobicoke                                          Northwest   

      Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0    43.806686 -79.194353               0  Fast Food Restaurant   
1    43.784535 -79.160497               0                   Bar   
2    43.763573 -79.188711               0           Pizza Place   
3    43.770992 -79.216917               0           Coffee Shop   
4    43.773136 -79.239476               0                Bakery   
..         ...        ...             ...                   ...   
95   43.643515 -79.577201               0              Pharmacy   
96   43.756303 -79.565963               0   Empanada Restaurant   
99   43.696319 -79.532242               0           Pizza Place   
101  43.739416 -79.588437               0         Grocery Store   
102  43.706748 -79.594054               0         Moving Target   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Yoga Studio     Electronics Store               Dog Run   
1                  Yoga Studio   Empanada Restaurant            Donut Shop   
2                          Spa        Breakfast Spot   Rental Car Location   
3            Indian Restaurant     Korean Restaurant           Yoga Studio   
4                         Bank      Hakka Restaurant    Athletics & Sports   
..                         ...                   ...                   ...   
95                Liquor Store                  Café        Shopping Plaza   
96           Electronics Store               Dog Run      Doner Restaurant   
99   Middle Eastern Restaurant           Coffee Shop    Chinese Restaurant   
101                   Pharmacy            Beer Store  Fast Food Restaurant   
102        Rental Car Location                   Bar             Drugstore   

    5th Most Common Venue  
0        Doner Restaurant  
1               Drugstore  
2          Medical Center  
3       Electronics Store  
4         Thai Restaurant  
..                    ...  
95             Beer Store  
96             Donut Shop  
99         Sandwich Place  
101        Sandwich Place  
102                 Diner  

[85 rows x 11 columns]

## Cluster 2:  Recreation

In [75]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1]

Postcode           Borough  \
14       M1V       Scarborough   
23       M2P        North York   
25       M3A        North York   
30       M3K        North York   
40       M4J         East York   
44       M4N   Central Toronto   
48       M4T   Central Toronto   
50       M4W  Downtown Toronto   
74       M6E              York   
90       M8X         Etobicoke   
98       M9N              York   
100      M9R         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
14   Agincourt North,L'Amoreaux East,Milliken,Steel...  43.815252 -79.284577   
23                                     York Mills West  43.752758 -79.400049   
25                                           Parkwoods  43.753259 -79.329656   
30                          CFB Toronto,Downsview East  43.737473 -79.464763   
40                                        East Toronto  43.685347 -79.338106   
44                                       Lawrence Park  43.728020 -79.388790   
48                          Moore Park,Summerhill East  43.689574 -79.383160   
50                                            Rosedale  43.679563 -79.377529   
74                                 Caledonia-Fairbanks  43.689026 -79.453512   
90         The Kingsway,Montgomery Road,Old Mill North  43.653654 -79.506944   
98                                              Weston  43.706876 -79.518188   
100  Kingsview Village,Martin Grove Gardens,Richvie...  43.688905 -79.554724   

     Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
14                1                  Park            Playground   
23                1                  Bank                  Park   
25                1  Fast Food Restaurant                  Park   
30                1               Airport                  Park   
40                1                  Park     Convenience Store   
44                1                  Park           Swim School   
48                1          Tennis Court                  Park   
50                1                  Park              Building   
74                1                  Park                Market   
90                1                 River                  Park   
98                1                  Park           Yoga Studio   
100               1           Pizza Place                  Park   

    3rd Most Common Venue        4th Most Common Venue  \
14            Yoga Studio            Electronics Store   
23      Convenience Store                  Yoga Studio   
25      Food & Drink Shop                  Yoga Studio   
30            Yoga Studio          Empanada Restaurant   
40            Yoga Studio          Empanada Restaurant   
44               Bus Line                  Yoga Studio   
48            Yoga Studio  Eastern European Restaurant   
50             Playground                        Trail   
74          Women's Store         Fast Food Restaurant   
90                   Pool  Eastern European Restaurant   
98    Empanada Restaurant             Doner Restaurant   
100     Mobile Phone Shop                     Bus Line   

           5th Most Common Venue  
14                       Dog Run  
23           Empanada Restaurant  
25             Electronics Store  
30              Doner Restaurant  
40              Doner Restaurant  
44   Eastern European Restaurant  
48                       Dog Run  
50                   Yoga Studio  
74             Electronics Store  
90                Discount Store  
98                    Donut Shop  
100                  Yoga Studio

## Cluster 3: Recreation (should have been clustered with cluster 2)

In [78]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2]

Postcode     Borough                                      Neighbourhood  \
91      M8Y   Etobicoke  Humber Bay,King's Mill Park,Kingsway Park Sout...   
97      M9M  North York                                    Emery,Humberlea   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
91  43.636258 -79.498509               2        Baseball Field   
97  43.724766 -79.532242               2        Baseball Field   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
91                  Pool           Yoga Studio     Electronics Store   
97           Yoga Studio   Empanada Restaurant            Donut Shop   

   5th Most Common Venue  
91      Doner Restaurant  
97             Drugstore